### ISSUE TITLE
Add pain.008.003.02 support

### ISSUE THREAD

**Issue by osopolar** [php-sepa-xml/php-sepa-xml#59]
Follow-up from #47:

> We just have to evaluate the specific changes in 008.03.02 and could then implement it. The biggest deal in doing this is to figure out what change between 02 and 03.

The official files [pain.008.003.02.xsd](http://www.ebics.de/fileadmin/unsecured/anlage3/anlage3_pain008/pain_schema/pain.008.003.02.xsd) and [pain.008.003.02.xml](http://www.ebics.de/fileadmin/unsecured/anlage3/anlage3_pain008/pain_ex/pain.008.003.02.xml) from ebics.de linked at http://www.sepazahlungsverkehr.de/technische-anforderungen/sepa-lastschrift are not available. I found a copy at [github/w2c/sepa-sdd-xml-generator](https://github.com/w2c/sepa-sdd-xml-generator/blob/master/validation_schemes/pain.008.003.02.xsd).

I attached a [diff between the xsd files]((https://github.com/php-sepa-xml/php-sepa-xml/files/1428872/diff-pain.008.002.02.xsd-pain.008.003.02.xsd.diff.txt)) (`diff pain.008.002.02.xsd pain.008.003.02.xsd -Naurb`).

As the comment in pain.008.003.02.xsd says, the file pain.008.002.02.xsd was modified with XMLSpy v2008 rel. 2; it implements IBAN Only transfers, COR1-Option and extends service level to external code Lists:

>Mit XMLSpy v2008 rel. 2 (http://www.altova.com) von Wenzel (SIZ Bonn) bearbeitet
>Version gemäß DFÜ-Abkommen Anlage 3, Version 2.7, gültig ab November 2013 mit Umsetzung von IBAN Only gemäß EPC SDD Core IG 7.0 bzw. SDD B2B IG 5.0 , zudem Einbau der COR1-Option durch Erweiterung Local Instrument und Erweiterung Service Level auf Externe Codeliste

IBAN Only is already supported by php-sepa-xml (FinInstnId > Othr > Id: NOTPROVIDED), so at least that seem to be defined in an earlier pain format.

```diff
--- pain.008.002.02.xsd	2017-10-30 18:18:24.000000000 -0300
+++ pain.008.003.02.xsd	2017-10-30 18:17:43.000000000 -0300
@@ -1,5 +1,8 @@
 <?xml version="1.0" encoding="UTF-8" standalone="no"?>
-<xs:schema xmlns="urn:iso:std:iso:20022:tech:xsd:pain.008.002.02" xmlns:xs="http://www.w3.org/2001/XMLSchema" targetNamespace="urn:iso:std:iso:20022:tech:xsd:pain.008.002.02" elementFormDefault="qualified">
+<!-- Mit XMLSpy v2008 rel. 2 (http://www.altova.com) von Wenzel (SIZ Bonn) bearbeitet -->
+<!-- Version gemäß DFÜ-Abkommen Anlage 3, Version 2.7, gültig ab November 2013 mit Umsetzung von IBAN Only gemäß EPC SDD Core IG 7.0 bzw. SDD B2B IG 5.0 , zudem Einbau der COR1-Option durch Erweiterung Local Instrument und Erweiterung Service Level auf Externe Codeliste-->
+<!-- Mit XMLSpy v2008 rel. 2 sp2 (http://www.altova.com) am 29.11.2012 von der SIZ GmbH bearbeitet -->
+<xs:schema xmlns="urn:iso:std:iso:20022:tech:xsd:pain.008.003.02" xmlns:xs="http://www.w3.org/2001/XMLSchema" targetNamespace="urn:iso:std:iso:20022:tech:xsd:pain.008.003.02" elementFormDefault="qualified">
 	<xs:element name="Document" type="Document"/>
 	<xs:complexType name="AccountIdentificationSEPA">
 		<xs:sequence>
@@ -68,9 +71,9 @@
 	<xs:simpleType name="BatchBookingIndicator">
 		<xs:restriction base="xs:boolean"/>
 	</xs:simpleType>
-	<xs:complexType name="BranchAndFinancialInstitutionIdentificationSEPA1">
+	<xs:complexType name="BranchAndFinancialInstitutionIdentificationSEPA3">
 		<xs:sequence>
-			<xs:element name="FinInstnId" type="FinancialInstitutionIdentificationSEPA1"/>
+			<xs:element name="FinInstnId" type="FinancialInstitutionIdentificationSEPA3"/>
 		</xs:sequence>
 	</xs:complexType>
 	<xs:complexType name="BranchAndFinancialInstitutionIdentificationSEPA2">
@@ -173,7 +176,7 @@
 					<xs:documentation>This data element may be present either at ‘Payment Information’ or at ‘Direct Debit Transaction Information’ level.</xs:documentation>
 				</xs:annotation>
 			</xs:element>
-			<xs:element name="DbtrAgt" type="BranchAndFinancialInstitutionIdentificationSEPA1"/>
+			<xs:element name="DbtrAgt" type="BranchAndFinancialInstitutionIdentificationSEPA3"/>
 			<xs:element name="Dbtr" type="PartyIdentificationSEPA2"/>
 			<xs:element name="DbtrAcct" type="CashAccountSEPA2"/>
 			<xs:element name="UltmtDbtr" type="PartyIdentificationSEPA1" minOccurs="0">
@@ -201,10 +204,10 @@
 			<xs:maxLength value="4"/>
 		</xs:restriction>
 	</xs:simpleType>
-	<xs:simpleType name="LocalInstrumentSEPACode">
+	<xs:simpleType name="ExternalLocalInstrument1Code">
 		<xs:restriction base="xs:string">
-			<xs:enumeration value="CORE"/>
-			<xs:enumeration value="B2B"/>
+			<xs:minLength value="1"/>
+			<xs:maxLength value="35"/>
 		</xs:restriction>
 	</xs:simpleType>
 	<xs:simpleType name="ExternalOrganisationIdentification1Code">
@@ -225,16 +228,30 @@
 			<xs:maxLength value="4"/>
 		</xs:restriction>
 	</xs:simpleType>
-	<xs:simpleType name="ServiceLevelSEPACode">
+	<xs:simpleType name="ExternalServiceLevel1Code">
 		<xs:restriction base="xs:string">
-			<xs:enumeration value="SEPA"/>
+			<xs:minLength value="1"/>
+			<xs:maxLength value="4"/>
 		</xs:restriction>
 	</xs:simpleType>
-	<xs:complexType name="FinancialInstitutionIdentificationSEPA1">
+	<xs:complexType name="FinancialInstitutionIdentificationSEPA3">
 		<xs:sequence>
+			<xs:choice>
 			<xs:element name="BIC" type="BICIdentifier"/>
+				<xs:element name="Othr" type="OthrIdentification"/>
+			</xs:choice>
 		</xs:sequence>
 	</xs:complexType>
+	<xs:complexType name="OthrIdentification">
+		<xs:sequence>
+			<xs:element name="Id" type="OthrIdentificationCode"/>
+		</xs:sequence>
+	</xs:complexType>
+	<xs:simpleType name="OthrIdentificationCode">
+		<xs:restriction base="xs:string">
+			<xs:enumeration value="NOTPROVIDED"/>
+		</xs:restriction>
+	</xs:simpleType>
 	<xs:complexType name="FinancialInstitutionIdentificationSEPA2">
 		<xs:sequence>
 			<xs:element name="Othr" type="RestrictedFinancialIdentificationSEPA"/>
@@ -297,7 +314,7 @@
 	</xs:simpleType>
 	<xs:complexType name="LocalInstrumentSEPA">
 		<xs:sequence>
-			<xs:element name="Cd" type="LocalInstrumentSEPACode"/>
+			<xs:element name="Cd" type="ExternalLocalInstrument1Code"/>
 		</xs:sequence>
 	</xs:complexType>
 	<xs:complexType name="MandateRelatedInformationSDD">
@@ -453,7 +470,7 @@
 			<xs:element name="ReqdColltnDt" type="ISODate"/>
 			<xs:element name="Cdtr" type="PartyIdentificationSEPA5"/>
 			<xs:element name="CdtrAcct" type="CashAccountSEPA1"/>
-			<xs:element name="CdtrAgt" type="BranchAndFinancialInstitutionIdentificationSEPA1"/>
+			<xs:element name="CdtrAgt" type="BranchAndFinancialInstitutionIdentificationSEPA3"/>
 			<xs:element name="UltmtCdtr" type="PartyIdentificationSEPA1" minOccurs="0">
 				<xs:annotation>
 					<xs:documentation>This data element may be present either at ‘Payment Information’ or at ‘Direct Debit Transaction Information’ level.</xs:documentation>
@@ -484,7 +501,7 @@
 			<xs:element name="SvcLvl" type="ServiceLevelSEPA"/>
 			<xs:element name="LclInstrm" type="LocalInstrumentSEPA">
 				<xs:annotation>
-					<xs:documentation>Only ‘B2B’ or 'CORE' is allowed. The mixing of Core Direct Debits and B2B Direct Debits is not allowed in the same message.</xs:documentation>
+					<xs:documentation>Only ‘B2B’, 'CORE' or 'COR1' is allowed. The mixing of different Local Instrument values is not allowed in the same message.</xs:documentation>
 				</xs:annotation>
 			</xs:element>
 			<xs:element name="SeqTp" type="SequenceType1Code">
@@ -569,7 +586,7 @@
 	</xs:simpleType>
 	<xs:complexType name="ServiceLevelSEPA">
 		<xs:sequence>
-			<xs:element name="Cd" type="ServiceLevelSEPACode"/>
+			<xs:element name="Cd" type="ExternalServiceLevel1Code"/>
 		</xs:sequence>
 	</xs:complexType>
 	<xs:complexType name="StructuredRemittanceInformationSEPA1">
```

**Comment by osopolar** [comment]
Maybe this Tryton issue helps too: https://bugs.tryton.org/issue4216

**Comment by osopolar** [comment]
I asked at info [at] ebics.de if they could provide the [missing pain.008.003.02.xsd file](http://www.ebics.de/fileadmin/unsecured/anlage3/anlage3_pain008/pain_schema/pain.008.003.02.xsd).

I got following answer:

> Sie suchen eine alte Fassung des SEPA Technical Validation Subsets (TVS) der Deutschen Kreditwirtschaft. Die aktuellen Dokumente (xsd's) finden  sie direkt unter http://www.ebics.de/spezifikation/dfue-abkommen-anlage-3-formatstandards/
>
> Anlage 3 ist in Kürze (19.11.2017) in der Version 3.1 gültig.
>
> Auf der gleichen Seite finden Sie unten einen Bereich "Alte Fassungen (Archiv)". Die gesuchte Datei wurde mit der Version 2.9 der Anlage 3 archiviert.
...

This means so far, that pain.008.003.02 will be outdated by 19.11.2017. [There is a new version 3.1](http://www.ebics.de/spezifikation/dfue-abkommen-anlage-3-formatstandards/).

As I am not very familiar with the versions of the standard and the implications of the changes, I asked for more information and got the following answer:

> Die ab 19.11.2017 gültige Fassung wird in Absprache mit ISO als "ISO-Variante" "001" geführt, da das TVS zwar gegenüber original ISO für den Zweck SEPA etwas reduziert ist, darüber validierte XML-Dateien jedoch voll ISO pain.008.001.02-kompatibel sind.
> Das TVS der DK hat aktuell den Dateinamen pain.008.001.02_GBIC2 und ist in dem zip-Container SEPA-Schema_V3.1_GBIC2 enthalten. Diesen also bitte öffnen. Bitte auch  die Texte zu dieser Datei beachten. Müsste eigentlich alles selbsterklärend sein.

As far as I understand this information, there is no need to implement pain.008.003.02. The new version (3.1) will be compatible with pain.008.001.02. The Technical Validation Subsets (TVS) of the  German Banking Industry Committee can be found as pain.008.001.02_GBIC2 in the zip archive [SEPAContainer Schema V3.1 GBIC 2.zip](http://www.ebics.de/fileadmin/unsecured/anlage3/anlage3_container/SEPAContainer_Schema_V3.1_GBIC_2.zip) at the above mentioned page.

**Comment by monofone** [comment]
Thanks for sharing your research

**Comment by christian-ehmig** [comment]
Although the new version (3.1) is supposed to be compatible with pain.008.001.02, we need to pass additional information for direct debit transfers described in the document. E.g. SEPA transfers from EU countries outside Germany need to pass the Debitor name, address and country within additional fields.
Specifically http://www.ebics.de/fileadmin/unsecured/anlage3/anlage3_spec/Anlage_3_Datenformate_V_3.1.pdf page 72, 2.2.2.10 Debtor
Fields PstlAdr, Ctry, AdrLine

Example:
```xml
<Dbtr>
  <Nm>Debtor Name</Nm>
  <PstlAdr>
  	<Ctry>AT</Ctry>
  	<AdrLine>Debtor Street</AdrLine>
  	<AdrLine>12345 Debtor City</AdrLine>
  </PstlAdr>
</Dbtr>
```
Are you going to add those fields in the near future?

**Comment by monofone** [comment]
Yes I will implement the changes, would be great if you can support by testing the changes.

**Comment by christian-ehmig** [comment]
Sure - I will gladly support you by testing.

**Comment by christian-ehmig** [comment]
Any news on this one?

**Comment by monofone** [comment]
in progress ...
	Any news on this one?

**Comment by monofone** [comment]
@christian-ehmig here are the news, if you can check on this one, would be nice

**Comment by christian-ehmig** [comment]
@monofone works perfectly in my case - tested with setting painFormat = pain.008.003.02, created a directDebit XML with about 300 transactions, country & address info - country & address lines were correctly set.

**Comment by monofone** [comment]
@christian-ehmig it took a while ... can you confirm that the master is working for you ? then I'll build a new release.

**Comment by christian-ehmig** [comment]
@monofone this time it took a while on my side, sry! Works perfectly (dev-master). Thanks a lot!

**Comment by monofone** [comment]
Released 1.3.0
